Various updates to handle HRD API changes

johnclary opened the pull request:
* https://github.com/cityofaustin/atd-data-tech/issues/23342

### Testing

1. Navigate to the root of this repo and checkout this branch.
2. In the **API Accessible Secrets** vault in 1pass, locate the item called **HRD Banner API** and save a copy of the `data.xml` file attached to it. Place the file in the root of this repo.
3. Create your `env_file` using the instructions in the env template. ⚠️ Be sure to use the Knack credentials in the **TEST** section of the 1pass item ⚠️
4. Execute the script by mounting your local repo inside the existing Docker image:

```shell
docker run -it --rm --env-file env_file \
    -v ${PWD}:/app \
    atddocker/atd-knack-banner:production ./update_employees.py
```

5. This may detect some updates to be made depending on the state of the user table in the test app.
6. Use the **Knack Builder - Primary Account** credentials to go to the [knack builder ](https://builder.knack.com/atd/test--hr--7-jul-2025--human-resources-manager/records/objects/object_5) and delete an active user record from the test app.
7. Run the ETL again and confirm it creates a user
8. Modify the `data.xml` file by removing an an entire user object from the data and changing the name of another user.  Run the ETL again and observe that the records are updated in Knack.

johnclary committed:
rework email handling

johnclary commented on update_employees.py:
The significant change we're handling is that the old HR API used to send the string `"no email"` when a user did not have an email address.  The new API value uses an empty string (`""`).  We are now coercing the empty string to `None`/`null`.

johnclary commented on update_employees.py:
```diff
@@ -118,6 +119,15 @@ def drop_empty_positions(records_hr, key="pidm"):
+        json_raw = wddx.loads(records_hr_unfiltered)
+        #  remove weird leading slashes from data contents
+        json_clean = json_raw[0].replace("//", "")
+        return json.loads(json_clean)
```
We lost VPN access to the API, so until that's fixed this hack makes it possible to save a copy of the response payload in a local file so that we can mock the API (ty Chia for this suggestion).

Todo: remove this before merging.

mddilley replied:
thank y'all! this made it really straightforward to test the different operations. 🙏

johnclary committed:
update ignores

johnclary commented on update_employees.py:
```diff
@@ -270,11 +285,12 @@ def build_payload(
                 # if any of the fields differ, add banner record to payload
                 if is_different(r_hr, r_knack):
                     payload.append(r_hr)
-                    result["updates"].append((r_hr[name_field]))
```
i removed these appends to `result` because they were falling out of sync with what records were actually updated due to filtering that happens later on in the script.

johnclary committed:
remove linebreak

johnclary committed:
randomize email on user create

johnclary commented on update_employees.py:
```diff
@@ -137,22 +147,19 @@ def get_employee_data():
-    :param name_field: name field in knack
-    :return: temporary placeholder email
+    Emails are required for Knack user records. When an employee doesn't have an email
+    address, we generate a unique address using the current timestamp
```
by using a random email address we are avoiding future email address collisions when generating email addresses. The main issue we're working around is that seasonal employees like crossing guards do not have email addresses, and they are often re-hired under a different employee ID. If you search the prod user table for the first name `Lanna` you will see an example of this. Prior to this change, we ran into a collision when trying to re-generate a different employee ID using email`first_name.last_name@austintexas.gov` pattern. This also avoids another edge case where we might generate a user's email address but in fact that email address actually matches a different employee.

the assumption here is that if the user does not have an email address then they do not need to login to the HR portal. if HR wants such an employee to be able to access the portal, they need to work with CTM to get the user a valid email address and to update that info in the upstream system serving this API.

mddilley replied:
Thanks for the background, and this change to keep CTM as the source of valid emails makes sense to me!

johnclary commented on update_employees.py:
```diff
@@ -344,27 +362,6 @@ def set_passwords(records, password_field):
     return
 
 
-def remove_empty_emails(payload, email_field, name_field):
```
as far as I can tell this isn't possible because we assign emails to new users who don't have them.

johnclary commented on update_employees.py:
```diff
@@ -380,9 +377,13 @@ def main():
     logging.info("Getting employee data from Banner...")
-    records_hr_banner = get_employee_data()
+    # records_hr_banner = get_employee_data()
+    records_hr_banner = get_employee_data_local()
```
todo: fix this line before merging

Charlie-Henry replied:
maybe make this a script parameter? like `--local`? could be useful in the future

johnclary replied:
now that we have VPN access fixed I'm going to remove this code. there's no reason why we can't test with that prod endpoint

johnclary commented on update_employees.py:
```diff
@@ -417,26 +418,23 @@ def main():
             app.record(data=record, method=method, obj=knack_obj)
         except requests.HTTPError as e:
             if e.response.status_code == 400:
-                if record["field_230"] == "Crossing Guard":
```
i was hitting an occasional type error here dealing with de-activated users. i think it's fine to log all this below with the rest of the errors

johnclary committed:
update env template

mddilley commented on update_employees.py:
```diff
@@ -270,11 +283,12 @@ def build_payload(
         # employee id number not in knack records
         if not exists_in_knack:
-
+            # hand empty emails
```
👀 tiny typo - thanks for the extra comments and logging in this code btw 🙏

chiaberry replied:
✋🏼

mddilley reviewed (approved):
I tested combinations of edits, additions, removals in the very slick test file. I see the expected updates in the test app and events logged. Thanks for patching up this unexpected change! 🙏 🚢

Charlie-Henry reviewed (approved):
Tested like a charm!

<img width="147" height="145" alt="Screenshot 2025-07-10 at 1 14 46 PM" src="https://github.com/user-attachments/assets/dc063f75-be12-4993-99cb-a4501f24ba7e" />

chiaberry reviewed (approved):
thank you! 🚢

johnclary committed:
various updates
- remove local json file option
- fix a bug where we werent properly handling
  when an incoming email address is empty
- improve logging
- cut the sleep time when generating emails  down
  to .001 seconds

johnclary commented on update_employees.py:
```diff
@@ -261,7 +265,11 @@ def build_payload(
                 r_hr["id"] = r_knack["id"]
+                # fill in missing email addresses from Knack - some employees (temp, seasonal) do not have email addresses
+                if not r_hr[email_field]["email"]:
+                    r_hr[email_field]["email"] = r_knack[email_field]["email"]
```
i uncovered a bug when running this locally just now (because we had a few hundred record updates come in from the prod API!). for existing records, we need to assign the Knack email address to the incoming record before we diff it. these are records that do not have an email address in the API data but do already exist in Knack (and therefore must have had an email address generated for them by our script).

i introduced a new flavor of this bug, but i think previously we may have been skipping updates for users with no email address (via the `remove_empty_emails()` function i removed). 

anyway—this is fixed!
